File: go/internal/execx/execx.go
package execx

import (
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)

type Runner interface {
	Run(cmd string, args ...string) (stdout string, exitCode int, err error)
}

type System struct{}

func (System) Run(cmd string, args ...string) (string, int, error) {
	c := exec.Command(cmd, args...)
	out, err := c.Output()
	exit := 0
	if err != nil {
		if ee, ok := err.(*exec.ExitError); ok {
			exit = ee.ExitCode()
		} else {
			return "", -1, err
		}
	}
	return string(out), exit, nil
}

type Fixture struct {
	Dir string
}

func (f Fixture) Run(cmd string, args ...string) (string, int, error) {
	path := filepath.Join(f.Dir, FixtureKey(cmd, args)+".txt")
	b, err := os.ReadFile(path)
	if err != nil {
		if os.IsNotExist(err) {
			return "", 1, nil
		}
		return "", -1, err
	}
	return string(b), 0, nil
}

func FixtureKey(cmd string, args []string) string {
	parts := []string{sanitize(cmd)}
	for _, a := range args {
		parts = append(parts, sanitize(a))
	}
	return strings.Join(parts, "__")
}

func sanitize(s string) string {
	r := strings.NewReplacer("/", "_", " ", "_")
	return r.Replace(s)
}

File: rust/src/execx.rs
use std::path::PathBuf;
use std::process::Command;

pub trait Runner {
    fn run(&self, cmd: &str, args: &[&str]) -> std::io::Result<(String, i32)>;
}

pub struct System;

impl Runner for System {
    fn run(&self, cmd: &str, args: &[&str]) -> std::io::Result<(String, i32)> {
        let output = Command::new(cmd).args(args).output()?;
        let stdout = String::from_utf8_lossy(&output.stdout).into_owned();
        let code = output.status.code().unwrap_or(-1);
        Ok((stdout, code))
    }
}

pub struct Fixture {
    pub dir: PathBuf,
}

impl Runner for Fixture {
    fn run(&self, cmd: &str, args: &[&str]) -> std::io::Result<(String, i32)> {
        let path = self.dir.join(format!("{}.txt", fixture_key(cmd, args)));
        match std::fs::read_to_string(&path) {
            Ok(s) => Ok((s, 0)),
            Err(e) if e.kind() == std::io::ErrorKind::NotFound => Ok((String::new(), 1)),
            Err(e) => Err(e),
        }
    }
}

pub fn fixture_key(cmd: &str, args: &[&str]) -> String {
    let mut parts: Vec<String> = vec![sanitize(cmd)];
    for a in args {
        parts.push(sanitize(a));
    }
    parts.join("__")
}

fn sanitize(s: &str) -> String {
    s.replace('/', "_").replace(' ', "_")
}
